cvmfsexec-osg-wrapper should use cvmfsexec where available (SOFTWARE-4950)

matyasselmeci opened the pull request:
- use cvmfsexec instead of mountrepo if possible (SOFTWARE-4950)
- Remove TACC-specific code
- dump cvmfsexec logs if we fail to mount a repo

matyasselmeci commented:
Tested on CHTC (where it properly falls back to `mountrepo`) and Frontera (where it does what it says on the can). Also removed TACC-specific code, which got moved to the CE's `slurm_local_attributes.sh` instead.

brianhlin commented on scripts/cvmfsexec-osg-wrapper:
```diff
@@ -114,30 +105,39 @@ if [[ -n $CVMFS_HTTP_PROXY ]]; then
+    command -v fusermount >& /dev/null || fail "Required command 'fusermount' not found"
 
-export CVMFS_BASE="$workdir"/cvmfsexec/dist/cvmfs
+    "$workdir"/cvmfsexec/umountrepo -a || :
```
I'm a little lost, what's this for exactly?

matyasselmeci replied:
In case there are leftover mounts from a previous run that never had time to unmount.

Thinking about it, we probably shouldn't do this unless our jobs are whole-machine jobs.

brianhlin replied:
At the very least, could you add a comment?

brianhlin commented on scripts/cvmfsexec-osg-wrapper:
```diff
@@ -114,30 +105,39 @@ if [[ -n $CVMFS_HTTP_PROXY ]]; then
-# If we have to use this script to mount CVMFS, we can't use images from there anyway.
-export ALLOW_NONCVMFS_IMAGES=true
+    # Mount repos
+    for repo in config-osg.opensciencegrid.org $CVMFSEXEC_REPOS; do
+        if ! "$workdir"/cvmfsexec/mountrepo "$repo"; then
+            echo "$repo mount failed, dumping logs:" >&2
+            tail -n 20 "$workdir"/cvmfsexec/log/"$repo".log >&2
+            fail "Unable to mount cvmfs repo $repo"
+        fi
+    done
```
Are there any instances where we'd be ok with a partial list of successfully mounted repos?

matyasselmeci replied:
I'd say by default no but I could add that as an option.

brianhlin replied:
Yeah, let's just add it later if we need it

brianhlin commented on scripts/cvmfsexec-osg-wrapper:
```diff
@@ -114,30 +105,39 @@ if [[ -n $CVMFS_HTTP_PROXY ]]; then
+    done
 
-echo "*** cvmfs available at $CVMFS_BASE" >&2
+    export CVMFS_BASE="$workdir"/cvmfsexec/dist/cvmfs
+    export GLIDEIN_SINGULARITY_BINDPATH="$CVMFS_BASE:/cvmfs"
+    echo "*** cvmfs available at $CVMFS_BASE" >&2
```
Isn't this backwards? This looks like the `cvmfsexec` path and we're in the block where the `cvmfsexec` smoke test failed

matyasselmeci replied:
Nope, look at CVMFS_BASE: it's no longer /cvmfs, but something under $workdir.

brianhlin replied:
I was fundamentally confused about the `cvmfsexec` vs `mountrepo` paths, though I think some comments saying that `mountrepo` can't bind mount to `/cvmfs` would've cleared things up a bit for me

brianhlin reviewed (commented):
I've got some questions but I may just have things backwards in my own head

matyasselmeci commented:
OK I think I addressed your comments. I also added a new variable, `UMOUNTREPO_UNMOUNT_ALL` which controls running `umountrepo -a`. Can you take another look?

brianhlin commented on scripts/cvmfsexec-osg-wrapper:
```diff
@@ -21,27 +21,29 @@
-    fi
-fi
-
+UMOUNTREPO_UNMOUNT_ALL=${UMOUNTREPO_UNMOUNT_ALL-}
```
```suggestion
UMOUNTREPO_UNMOUNT_ALL=${UMOUNTREPO_UNMOUNT_ALL:-}
```

Are you using param expansion to explicitly set this to the empty string if unset?

matyasselmeci replied:
Yep which is why `:-` is redundant.

brianhlin replied:
I'm having trouble finding ${param-} in the bash manual, what's it do?

matyasselmeci replied:
Should be under parameter expansion.
- `${foo:-X}` means 'use X if $foo is undefined or empty'
- `${foo-X}` means 'use X if $foo is undefined'

The other fancy expansions `${foo:=X}`, `${foo:?X}`, `${foo:+X}` also have colonless versions that work similarly.

brianhlin replied:
Ah, I see it now in a sentence that's hidden away. Do they expect me to actually read through the manual and not just grep through it???

brianhlin commented on scripts/cvmfsexec-osg-wrapper:
```diff
@@ -114,30 +114,62 @@ if [[ -n $CVMFS_HTTP_PROXY ]]; then
-# If we have to use this script to mount CVMFS, we can't use images from there anyway.
-export ALLOW_NONCVMFS_IMAGES=true
+    # Note: do not use 'exec'; cleanup won't run
+    "$workdir"/cvmfsexec/cvmfsexec -N $CVMFSEXEC_REPOS -- "$@"
```
```suggestion
    "$workdir"/cvmfsexec/cvmfsexec -N "$CVMFSEXEC_REPOS" -- "$@"
```

This comes from the user, right?

matyasselmeci replied:
Correct and it _should_ be expanded into separate arguments. It's not eval'ed so we don't have to worry about "little Bobby Tables" issues, although I could sanitize the variable names for safety's sake...

brianhlin reviewed (approved):
Some comments but mostly LGTM. Pre-approving
